Unable to handle plain text

Server resource:

```java
@GET
@Path("plaintext")
public String retrievePlainText() {
 return "This is plain text.";
}
```

Client service:

```java
@GET
@Path("plaintext")
public void retrievePlainText(MethodCallback<String> callback);
```
Service call:

```java
myService.retrievePlainText(myCallback);
```

Client output:

```text
[INFO]  - Sending http request: GET http://127.0.0.1:8888/rest/plaintext ,timeout:0
[INFO]  - Received http response for request: GET http://127.0.0.1:8888/rest/plaintext
This is plain text.

[ERROR] - Could not parse response: org.fusesource.restygwt.client.ResponseFormatException: Response was NOT a valid JSON document
```

Add annotation to the client service, call the service again.

```java
@Consumes(MediaType.TEXT_PLAIN)
```

Client output:

```text
[INFO]  - Sending http request: GET http://127.0.0.1:8888/rest/plaintext ,timeout:0
```

Server output:

```text
[WARN] 415 - GET /rest/plaintext (127.0.0.1) 
   Request headers
      ...
      Accept: application/json
      ...
   Response headers
      Content-Type: text/html
      ...
```

--- a/restygwt/src/main/java/org/fusesource/restygwt/RestyGWT.gwt.xml
+++ b/restygwt/src/main/java/org/fusesource/restygwt/RestyGWT.gwt.xml
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <!--
 
-    Copyright (C) 2009-2012 the original author or authors.
+    Copyright (C) 2009-2015 the original author or authors.
     See the notice.md file distributed with this work for additional
     information regarding copyright ownership.
 
@@ -44,10 +44,21 @@
 
   <define-configuration-property name="org.fusesource.restygwt.jsontypeidresolver" is-multi-valued="true"/>
   
-  <define-configuration-property name="org.fusesource.restygwt.annotationresolver"
-        is-multi-valued="true"/>
+  <define-configuration-property name="org.fusesource.restygwt.annotationresolver" is-multi-valued="true"/>
 
   <define-configuration-property name="org.fusesource.restygwt.restyjsonserializergenerator" is-multi-valued="true"/>
+
+<!-- /gwt-user/core/src/com/google/gwt/user/RemoteService.gwt.xml -->
+
+  <!--
+    Declare a property to determine whether autodetection of text/* with @Consumes and @Produces annotations should be used.
+  -->
+  <define-property name="restygwt.autodetect.plaintText" values="true,false" />
+  
+  <!--
+    Default no autodetection for text/*, for backward compatibility.
+  -->
+  <set-property name="restygwt.autodetect.plaintText" value="false" />
 
   <source path="client"/>
   <source path="example/client"/>

--- a/restygwt/src/main/java/org/fusesource/restygwt/RestyGWT.gwt.xml
+++ b/restygwt/src/main/java/org/fusesource/restygwt/RestyGWT.gwt.xml
@@ -53,12 +53,12 @@
   <!--
     Declare a property to determine whether autodetection of text/* with @Consumes and @Produces annotations should be used.
   -->
-  <define-property name="restygwt.autodetect.plaintText" values="true,false" />
+  <define-property name="restygwt.autodetect.plainText" values="true,false" />
   
   <!--
     Default no autodetection for text/*, for backward compatibility.
   -->
-  <set-property name="restygwt.autodetect.plaintText" value="false" />
+  <set-property name="restygwt.autodetect.plainText" value="false" />
 
   <source path="client"/>
   <source path="example/client"/>

--- a/restygwt/src/main/java/org/fusesource/restygwt/client/Method.java
+++ b/restygwt/src/main/java/org/fusesource/restygwt/client/Method.java
@@ -1,5 +1,5 @@
 /**
- * Copyright (C) 2009-2012 the original author or authors.
+ * Copyright (C) 2009-2015 the original author or authors.
  * See the notice.md file distributed with this work for additional
  * information regarding copyright ownership.
  *
@@ -219,8 +219,12 @@ public class Method {
         }
         return this.logger;
     }
-    
+
     public Object send(final TextCallback callback) {
+        return send((MethodCallback<String>) callback);
+    }
+
+    public Object send(final MethodCallback<String> callback) {
         defaultAcceptType(Resource.CONTENT_TYPE_TEXT);
         try {
             return send(new AbstractRequestCallback<String>(this, callback) {

--- a/restygwt/src/main/java/org/fusesource/restygwt/rebind/RestServiceClassCreator.java
+++ b/restygwt/src/main/java/org/fusesource/restygwt/rebind/RestServiceClassCreator.java
@@ -388,6 +388,35 @@ public class RestServiceClassCreator extends BaseSourceCreator {
                " : com.google.gwt.http.client.URL.encodePathSegment(" + expr + ")))+\"");
     }
 
+    void writeOptions(Options options, Options classOptions) {
+        // configure the dispatcher
+        if (options != null && options.dispatcher() != Dispatcher.class) {
+            // use the dispatcher configured for the method.
+            p("__method.setDispatcher(" + options.dispatcher().getName() + ".INSTANCE);");
+        } else {
+            // use the default dispatcher configured for the service..
+            p("__method.setDispatcher(this.dispatcher);");
+        }
+
+        // configure the expected statuses..
+        if (options != null && options.expect().length != 0) {
+            // Using method level defined expected status
+            p("__method.expect(" + join(options.expect(), ", ") + ");");
+        } else if (classOptions != null && classOptions.expect().length != 0) {
+            // Using class level defined expected status
+            p("__method.expect(" + join(classOptions.expect(), ", ") + ");");
+        }
+
+        // configure the timeout
+        if (options != null && options.timeout() >= 0) {
+            // Using method level defined value
+            p("__method.timeout(" + options.timeout() + ");");
+        } else if (classOptions != null && classOptions.timeout() >= 0) {
+            // Using class level defined value
+            p("__method.timeout(" + classOptions.timeout() + ");");
+        }
+    }
+
     private void writeMethodImpl(JMethod method, Options classOptions) throws UnableToCompleteException {
         boolean returnRequest = false;
         if (method.getReturnType() != JPrimitiveType.VOID) {
@@ -541,32 +570,7 @@ public class RestServiceClassCreator extends BaseSourceCreator {
                 }
             }
 
-            // configure the dispatcher
-            if( options!=null && options.dispatcher()!=Dispatcher.class ) {
-                // use the dispatcher configured for the method.
-                p("__method.setDispatcher("+options.dispatcher().getName()+".INSTANCE);");
-            } else {
-                // use the default dispatcher configured for the service..
-                p("__method.setDispatcher(this.dispatcher);");
-            }
-
-            // configure the expected statuses..
-            if( options!=null && options.expect().length!=0 ) {
-                // Using method level defined expected status
-                p("__method.expect("+join(options.expect(), ", ")+");");
-            } else if( classOptions!=null && classOptions.expect().length!=0 ) {
-                // Using class level defined expected status
-                p("__method.expect("+join(classOptions.expect(), ", ")+");");
-            }
-
-            // configure the timeout
-            if( options!=null && options.timeout() >= 0 ) {
-                // Using method level defined value
-                p("__method.timeout("+options.timeout()+");");
-            } else if( classOptions!=null && classOptions.timeout() >= 0 ) {
-                // Using class level defined value
-                p("__method.timeout("+classOptions.timeout()+");");
-            }
+            writeOptions(options, classOptions);
 
             if(jsonpAnnotation == null) {
                 Produces producesAnnotation = findAnnotationOnMethodOrEnclosingType(method, Produces.class);

--- a/restygwt/src/main/java/org/fusesource/restygwt/rebind/RestServiceClassCreator.java
+++ b/restygwt/src/main/java/org/fusesource/restygwt/rebind/RestServiceClassCreator.java
@@ -79,6 +79,7 @@ import com.google.gwt.core.client.JsArrayString;
 import com.google.gwt.core.ext.GeneratorContext;
 import com.google.gwt.core.ext.TreeLogger;
 import com.google.gwt.core.ext.UnableToCompleteException;
+import com.google.gwt.core.ext.typeinfo.HasAnnotations;
 import com.google.gwt.core.ext.typeinfo.JClassType;
 import com.google.gwt.core.ext.typeinfo.JGenericType;
 import com.google.gwt.core.ext.typeinfo.JMethod;
@@ -229,16 +230,7 @@ public class RestServiceClassCreator extends BaseSourceCreator {
         this.QUERY_PARAM_LIST_TYPES.add(find(Set.class, getLogger(), context));
 		this.REST_SERVICE_TYPE = find(RestService.class, getLogger(), context);
 
-        String path = null;
-        Path pathAnnotation = getAnnotation(source, Path.class);
-        if (pathAnnotation != null) {
-            path = pathAnnotation.value();
-        }
-
-        RemoteServiceRelativePath relativePath = getAnnotation(source, RemoteServiceRelativePath.class);
-        if (relativePath != null) {
-            path = relativePath.value();
-        }
+        String path = getPathFromSource(source);
 
         p("private " + RESOURCE_CLASS + " resource = null;");
         p();
@@ -300,6 +292,22 @@ public class RestServiceClassCreator extends BaseSourceCreator {
         	else
                 writeMethodImpl(method, options);
         }
+    }
+
+    private static String getPathFromSource(HasAnnotations annotatedType) {
+        String path = null;
+
+        Path pathAnnotation = getAnnotation(annotatedType, Path.class);
+        if (pathAnnotation != null) {
+            path = pathAnnotation.value();
+        }
+
+        RemoteServiceRelativePath relativePath = getAnnotation(annotatedType, RemoteServiceRelativePath.class);
+        if (relativePath != null) {
+            path = relativePath.value();
+        }
+
+        return path;
     }
 
     private String quote(String path) {

--- a/restygwt/src/main/java/org/fusesource/restygwt/rebind/RestServiceClassCreator.java
+++ b/restygwt/src/main/java/org/fusesource/restygwt/rebind/RestServiceClassCreator.java
@@ -76,7 +76,10 @@ import com.google.gwt.core.client.JsArrayBoolean;
 import com.google.gwt.core.client.JsArrayInteger;
 import com.google.gwt.core.client.JsArrayNumber;
 import com.google.gwt.core.client.JsArrayString;
+import com.google.gwt.core.ext.BadPropertyValueException;
 import com.google.gwt.core.ext.GeneratorContext;
+import com.google.gwt.core.ext.PropertyOracle;
+import com.google.gwt.core.ext.SelectionProperty;
 import com.google.gwt.core.ext.TreeLogger;
 import com.google.gwt.core.ext.UnableToCompleteException;
 import com.google.gwt.core.ext.typeinfo.HasAnnotations;
@@ -97,6 +100,7 @@ import com.google.gwt.jsonp.client.JsonpRequest;
 import com.google.gwt.user.client.rpc.RemoteServiceRelativePath;
 import com.google.gwt.user.rebind.ClassSourceFileComposerFactory;
 import com.google.gwt.xml.client.Document;
+
 /**
  *
  * @author <a href="http://hiramchirino.com">Hiram Chirino</a>
@@ -108,6 +112,8 @@ import com.google.gwt.xml.client.Document;
 public class RestServiceClassCreator extends BaseSourceCreator {
 
     private static final String REST_SERVICE_PROXY_SUFFIX = "_Generated_RestServiceProxy_";
+
+    private static final String PLAIN_TEXT_AUTODETECTION_CONFIGURATION_PROPERTY_NAME = "restygwt.autodetect.plaintText";
 
     private static final String METHOD_CLASS = Method.class.getName();
     private static final String RESOURCE_CLASS = Resource.class.getName();
@@ -166,6 +172,8 @@ public class RestServiceClassCreator extends BaseSourceCreator {
     private JClassType REST_SERVICE_TYPE;
     private JsonEncoderDecoderInstanceLocator locator;
 
+    private boolean autodetectTypeForStrings;
+
     public RestServiceClassCreator(TreeLogger logger, GeneratorContext context, JClassType source) {
         super(logger, context, source, REST_SERVICE_PROXY_SUFFIX);
     }
@@ -205,6 +213,8 @@ public class RestServiceClassCreator extends BaseSourceCreator {
             getLogger().log(ERROR, "Type is not an interface.");
             throw new UnableToCompleteException();
         }
+
+        autodetectTypeForStrings = shouldAutodetectPlainTextForStrings(getLogger(), context.getPropertyOracle());
 
         locator = new JsonEncoderDecoderInstanceLocator(context, getLogger());
 
@@ -292,6 +302,20 @@ public class RestServiceClassCreator extends BaseSourceCreator {
         	else
                 writeMethodImpl(method, options);
         }
+    }
+
+    /**
+     * Returns {@code true} if plain text autodetection for strings should be used.
+     */
+    static boolean shouldAutodetectPlainTextForStrings(TreeLogger logger, PropertyOracle propertyOracle) {
+        try {
+            SelectionProperty prop = propertyOracle.getSelectionProperty(logger, PLAIN_TEXT_AUTODETECTION_CONFIGURATION_PROPERTY_NAME);
+            String propVal = prop.getCurrentValue();
+            return Boolean.parseBoolean(propVal);
+        } catch (BadPropertyValueException e) {
+            // use default "false"
+        }
+        return false;
     }
 
     private static String getPathFromSource(HasAnnotations annotatedType) {
@@ -572,22 +596,28 @@ public class RestServiceClassCreator extends BaseSourceCreator {
 
             writeOptions(options, classOptions);
 
-            if(jsonpAnnotation == null) {
+            if (jsonpAnnotation == null) {
+                final String acceptHeader;
                 Produces producesAnnotation = findAnnotationOnMethodOrEnclosingType(method, Produces.class);
                 if (producesAnnotation != null) {
-                    p("__method.header(" + RESOURCE_CLASS + ".HEADER_ACCEPT, "+wrap(producesAnnotation.value()[0])+");");
+                    // Do not use autodetection, if accept type already set
+                    if (acceptTypeBuiltIn == null && autodetectTypeForStrings && producesAnnotation.value()[0].startsWith("text/")) {
+                        acceptTypeBuiltIn = "CONTENT_TYPE_TEXT";
+                    }
+                    acceptHeader = wrap(producesAnnotation.value()[0]);
                 } else {
-                    // set the default accept header....
+                    // set the default accept header value ...
                     if (acceptTypeBuiltIn != null) {
-                        p("__method.header(" + RESOURCE_CLASS + ".HEADER_ACCEPT, " + RESOURCE_CLASS + "." + acceptTypeBuiltIn + ");");
+                        acceptHeader = RESOURCE_CLASS + "." + acceptTypeBuiltIn;
                     } else {
-                        p("__method.header(" + RESOURCE_CLASS + ".HEADER_ACCEPT, " + RESOURCE_CLASS + ".CONTENT_TYPE_JSON);");
+                        acceptHeader = RESOURCE_CLASS + ".CONTENT_TYPE_JSON";
                     }
                 }
+                p("__method.header(" + RESOURCE_CLASS + ".HEADER_ACCEPT, " + acceptHeader + ");");
 
                 Consumes consumesAnnotation = findAnnotationOnMethodOrEnclosingType(method, Consumes.class);
                 if (consumesAnnotation != null) {
-                    p("__method.header(" + RESOURCE_CLASS + ".HEADER_CONTENT_TYPE, "+wrap(consumesAnnotation.value()[0])+");");
+                    p("__method.header(" + RESOURCE_CLASS + ".HEADER_CONTENT_TYPE, " + wrap(consumesAnnotation.value()[0]) + ");");
                 }
 
                 // and set the explicit headers now (could override the accept header)

--- a/restygwt/src/main/java/org/fusesource/restygwt/rebind/RestServiceClassCreator.java
+++ b/restygwt/src/main/java/org/fusesource/restygwt/rebind/RestServiceClassCreator.java
@@ -115,7 +115,7 @@ public class RestServiceClassCreator extends BaseSourceCreator {
 
     private static final String REST_SERVICE_PROXY_SUFFIX = "_Generated_RestServiceProxy_";
 
-    private static final String PLAIN_TEXT_AUTODETECTION_CONFIGURATION_PROPERTY_NAME = "restygwt.autodetect.plaintText";
+    private static final String PLAIN_TEXT_AUTODETECTION_CONFIGURATION_PROPERTY_NAME = "restygwt.autodetect.plainText";
 
     private static final String METHOD_CLASS = Method.class.getName();
     private static final String RESOURCE_CLASS = Resource.class.getName();

--- a/restygwt/src/main/java/org/fusesource/restygwt/rebind/RestServiceClassCreator.java
+++ b/restygwt/src/main/java/org/fusesource/restygwt/rebind/RestServiceClassCreator.java
@@ -46,6 +46,7 @@ import javax.ws.rs.Path;
 import javax.ws.rs.PathParam;
 import javax.ws.rs.Produces;
 import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.MediaType;
 
 import org.fusesource.restygwt.client.AbstractAsyncCallback;
 import org.fusesource.restygwt.client.AbstractRequestCallback;
@@ -95,6 +96,7 @@ import com.google.gwt.http.client.RequestException;
 import com.google.gwt.json.client.JSONArray;
 import com.google.gwt.json.client.JSONObject;
 import com.google.gwt.json.client.JSONParser;
+import com.google.gwt.json.client.JSONString;
 import com.google.gwt.json.client.JSONValue;
 import com.google.gwt.jsonp.client.JsonpRequest;
 import com.google.gwt.user.client.rpc.RemoteServiceRelativePath;
@@ -125,6 +127,7 @@ public class RestServiceClassCreator extends BaseSourceCreator {
     private static final String JSON_ARRAY_CLASS = JSONArray.class.getName();
     private static final String JSON_OBJECT_CLASS = JSONObject.class.getName();
     private static final String JSON_VALUE_CLASS = JSONValue.class.getName();
+    private static final String JSON_STRING_CLASS = JSONString.class.getName();
     private static final String REQUEST_EXCEPTION_CLASS = RequestException.class.getName();
     private static final String RESPONSE_FORMAT_EXCEPTION_CLASS = ResponseFormatException.class.getName();
     private static final String JSONP_METHOD_CLASS = JsonpMethod.class.getName();
@@ -528,8 +531,7 @@ public class RestServiceClassCreator extends BaseSourceCreator {
                 }
 
                 if (!formParams.isEmpty()) {
-                    getLogger().log(ERROR, "You can not have both @FormParam parameters and a content parameter: " +
-                                                method.getReadableDeclaration());
+                    getLogger().log(ERROR, "You can not have both @FormParam parameters and a content parameter: " + method.getReadableDeclaration());
                     throw new UnableToCompleteException();
                 }
 
@@ -596,8 +598,10 @@ public class RestServiceClassCreator extends BaseSourceCreator {
 
             writeOptions(options, classOptions);
 
-            if (jsonpAnnotation == null) {
-                final String acceptHeader;
+            String contentTypeHeaderValue = null;
+
+            if(jsonpAnnotation == null) {
+                final String acceptHeader; 
                 Produces producesAnnotation = findAnnotationOnMethodOrEnclosingType(method, Produces.class);
                 if (producesAnnotation != null) {
                     // Do not use autodetection, if accept type already set
@@ -617,8 +621,9 @@ public class RestServiceClassCreator extends BaseSourceCreator {
 
                 Consumes consumesAnnotation = findAnnotationOnMethodOrEnclosingType(method, Consumes.class);
                 if (consumesAnnotation != null) {
-                    p("__method.header(" + RESOURCE_CLASS + ".HEADER_CONTENT_TYPE, " + wrap(consumesAnnotation.value()[0]) + ");");
-                }
+                    contentTypeHeaderValue = consumesAnnotation.value()[0];
+                    p("__method.header(" + RESOURCE_CLASS + ".HEADER_CONTENT_TYPE, " + wrap(contentTypeHeaderValue) + ");");
+               }
 
                 // and set the explicit headers now (could override the accept header)
                 for (Map.Entry<String, JParameter> entry : headerParams.entrySet()) {
@@ -651,7 +656,11 @@ public class RestServiceClassCreator extends BaseSourceCreator {
 
             if (contentArg != null) {
                 if (contentArg.getType() == STRING_TYPE) {
-                    p("__method.text(" + contentArg.getName() + ");");
+                    if (autodetectTypeForStrings && MediaType.APPLICATION_JSON.equals(contentTypeHeaderValue)) {
+                        p("__method.json(new " + JSON_STRING_CLASS + "(" + contentArg.getName() + "));");
+                    } else {
+                        p("__method.text(" + contentArg.getName() + ");");
+                    }
                 } else if (contentArg.getType() == JSON_VALUE_TYPE) {
                     p("__method.json(" + contentArg.getName() + ");");
                 } else if (contentArg.getType().isClass() != null &&

--- a/restygwt/src/test/java/org/fusesource/restygwt/AutodetectPlainTextStringEncoderDecoderTestGwt.gwt.xml
+++ b/restygwt/src/test/java/org/fusesource/restygwt/AutodetectPlainTextStringEncoderDecoderTestGwt.gwt.xml
@@ -1,0 +1,26 @@
+<!--
+
+    Copyright (C) 2009-2015 the original author or authors.
+    See the notice.md file distributed with this work for additional
+    information regarding copyright ownership.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+-->
+
+<module>
+    <inherits name='org.fusesource.restygwt.StringEncoderDecoderTestGwt' />
+
+    <!-- <set-property name="restygwt.autodetect.plaintText" value="true" /> -->
+
+</module>

--- a/restygwt/src/test/java/org/fusesource/restygwt/AutodetectPlainTextStringEncoderDecoderTestGwt.gwt.xml
+++ b/restygwt/src/test/java/org/fusesource/restygwt/AutodetectPlainTextStringEncoderDecoderTestGwt.gwt.xml
@@ -21,6 +21,6 @@
 <module>
     <inherits name='org.fusesource.restygwt.StringEncoderDecoderTestGwt' />
 
-    <!-- <set-property name="restygwt.autodetect.plaintText" value="true" /> -->
+    <set-property name="restygwt.autodetect.plaintText" value="true" />
 
 </module>

--- a/restygwt/src/test/java/org/fusesource/restygwt/AutodetectPlainTextStringEncoderDecoderTestGwt.gwt.xml
+++ b/restygwt/src/test/java/org/fusesource/restygwt/AutodetectPlainTextStringEncoderDecoderTestGwt.gwt.xml
@@ -21,6 +21,6 @@
 <module>
     <inherits name='org.fusesource.restygwt.StringEncoderDecoderTestGwt' />
 
-    <set-property name="restygwt.autodetect.plaintText" value="true" />
+    <set-property name="restygwt.autodetect.plainText" value="true" />
 
 </module>

--- a/restygwt/src/test/java/org/fusesource/restygwt/GwtCompleteTestSuite.java
+++ b/restygwt/src/test/java/org/fusesource/restygwt/GwtCompleteTestSuite.java
@@ -45,7 +45,9 @@ import org.fusesource.restygwt.client.cache.VolatileQueueableCacheStorageTestGwt
 import org.fusesource.restygwt.client.codec.EncoderDecoderTestGwt;
 import org.fusesource.restygwt.client.codec.InnerClassesEncoderDecoderTestGwt;
 import org.fusesource.restygwt.client.codec.MapInRestServiceEncoderDecoderTestGwt;
+import org.fusesource.restygwt.client.complex.AutodetectPlainTextStringEncoderDecoderTestGwt;
 import org.fusesource.restygwt.client.complex.JsonTypeIdResolver;
+import org.fusesource.restygwt.client.complex.StringEncoderDecoderTestGwt;
 
 import com.google.gwt.junit.tools.GWTTestSuite;
 
@@ -105,6 +107,9 @@ public class GwtCompleteTestSuite extends TestCase {
         suite.addTestSuite(ConfiguredServiceTestGwt.class);
 
         suite.addTestSuite(JsonpTestGwt.class);
+
+        suite.addTestSuite(StringEncoderDecoderTestGwt.class);
+        suite.addTestSuite(AutodetectPlainTextStringEncoderDecoderTestGwt.class);
 
         return suite;
     }

--- a/restygwt/src/test/java/org/fusesource/restygwt/StringEncoderDecoderTestGwt.gwt.xml
+++ b/restygwt/src/test/java/org/fusesource/restygwt/StringEncoderDecoderTestGwt.gwt.xml
@@ -1,0 +1,29 @@
+<!--
+
+    Copyright (C) 2009-2015 the original author or authors.
+    See the notice.md file distributed with this work for additional
+    information regarding copyright ownership.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+-->
+
+<module>
+    <inherits name='com.google.gwt.user.User' />
+    <inherits name='com.google.gwt.logging.Logging'/>
+    <inherits name='org.fusesource.restygwt.RestyGWT'/>
+
+    <servlet path='/strings/*' class='org.fusesource.restygwt.server.complex.StringEncoderDecoderServlet' />
+
+    <source path='client'/>
+</module>

--- a/restygwt/src/test/java/org/fusesource/restygwt/client/complex/AutodetectPlainTextStringEncoderDecoderTestGwt.java
+++ b/restygwt/src/test/java/org/fusesource/restygwt/client/complex/AutodetectPlainTextStringEncoderDecoderTestGwt.java
@@ -1,0 +1,56 @@
+/**
+ * Copyright (C) 2009-2015 the original author or authors.
+ * See the notice.md file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.fusesource.restygwt.client.complex;
+
+import org.fusesource.restygwt.client.Method;
+import org.fusesource.restygwt.client.MethodCallback;
+
+import com.google.gwt.core.client.GWT;
+
+public class AutodetectPlainTextStringEncoderDecoderTestGwt extends StringEncoderDecoderTestGwt {
+
+    @Override
+    public String getModuleName() {
+        return "org.fusesource.restygwt.AutodetectPlainTextStringEncoderDecoderTestGwt";
+    }
+
+    /**
+     * Changed behaviour to only success through "onSuccess".
+     */
+    @Override
+    public void testPlainTextStringWithMethodCallback() {
+        StringsAsync strings = GWT.create(StringsAsync.class);
+
+        delayTestFinish(10000);
+
+        strings.getAsPlainText(new MethodCallback<String>() {
+            @Override
+            public void onSuccess(Method method, String response) {
+                assertEquals("String as plain text", response);
+                finishTest();
+            }
+
+            @Override
+            public void onFailure(Method method, Throwable exception) {
+                fail(exception.getMessage());
+            }
+        });
+    }
+
+}

--- a/restygwt/src/test/java/org/fusesource/restygwt/client/complex/AutodetectPlainTextStringEncoderDecoderTestGwt.java
+++ b/restygwt/src/test/java/org/fusesource/restygwt/client/complex/AutodetectPlainTextStringEncoderDecoderTestGwt.java
@@ -53,4 +53,23 @@ public class AutodetectPlainTextStringEncoderDecoderTestGwt extends StringEncode
         });
     }
 
+    @Override
+    public void testSendJsonString() {
+        StringsAsync strings = GWT.create(StringsAsync.class);
+
+        delayTestFinish(10000);
+
+        strings.setAsJson("\"Json String?\"", new MethodCallback<Void>() {
+            @Override
+            public void onSuccess(Method method, Void response) {
+                finishTest();
+            }
+
+            @Override
+            public void onFailure(Method method, Throwable exception) {
+                fail();
+            }
+        });
+    }
+
 }

--- a/restygwt/src/test/java/org/fusesource/restygwt/client/complex/StringEncoderDecoderTestGwt.java
+++ b/restygwt/src/test/java/org/fusesource/restygwt/client/complex/StringEncoderDecoderTestGwt.java
@@ -1,0 +1,132 @@
+/**
+ * Copyright (C) 2009-2015 the original author or authors.
+ * See the notice.md file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.fusesource.restygwt.client.complex;
+
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+
+import org.fusesource.restygwt.client.Method;
+import org.fusesource.restygwt.client.MethodCallback;
+import org.fusesource.restygwt.client.RestService;
+import org.fusesource.restygwt.client.TextCallback;
+
+import com.google.gwt.core.client.GWT;
+import com.google.gwt.json.client.JSONException;
+import com.google.gwt.junit.client.GWTTestCase;
+
+public class StringEncoderDecoderTestGwt extends GWTTestCase {
+
+    @Override
+    public String getModuleName() {
+        return "org.fusesource.restygwt.StringEncoderDecoderTestGwt";
+    }
+
+    @Path("/strings")
+    public static interface Strings {
+        @GET
+        @Produces(MediaType.APPLICATION_JSON)
+        String getAsJson();
+
+        @GET
+        @Produces(MediaType.TEXT_PLAIN)
+        String getAsPlainText();
+    }
+
+    @Path("/strings")
+    public static interface StringsAsync extends RestService {
+        @GET
+        @Produces(MediaType.APPLICATION_JSON)
+        void getAsJson(MethodCallback<String> callback);
+
+        @GET
+        @Produces(MediaType.TEXT_PLAIN)
+        void getAsPlainText(MethodCallback<String> callback);
+
+        @GET
+        @Produces(MediaType.TEXT_PLAIN)
+        void getAsPlainText(TextCallback callback);
+    }
+
+    public void testJsonString() {
+        StringsAsync strings = GWT.create(StringsAsync.class);
+
+        delayTestFinish(10000);
+
+        strings.getAsJson(new MethodCallback<String>() {
+            @Override
+            public void onSuccess(Method method, String response) {
+                assertEquals("String as Json", response);
+                finishTest();
+            }
+
+            @Override
+            public void onFailure(Method method, Throwable exception) {
+                fail(exception.getMessage());
+            }
+        });
+    }
+
+    public void testPlainTextStringWithTextCallback() {
+        StringsAsync strings = GWT.create(StringsAsync.class);
+
+        delayTestFinish(10000);
+
+        strings.getAsPlainText(new TextCallback() {
+            @Override
+            public void onSuccess(Method method, String response) {
+                assertEquals("String as plain text", response);
+                finishTest();
+            }
+
+            @Override
+            public void onFailure(Method method, Throwable exception) {
+                fail(exception.getMessage());
+            }
+        });
+    }
+
+    /**
+     * Test method only success through "onFailure" with restygwt <= 2.0.3 or plain text autodetection set to false (default)
+     */
+    public void testPlainTextStringWithMethodCallback() {
+        StringsAsync strings = GWT.create(StringsAsync.class);
+
+        delayTestFinish(10000);
+
+        strings.getAsPlainText(new MethodCallback<String>() {
+            @Override
+            public void onSuccess(Method method, String response) {
+                fail();
+            }
+
+            @Override
+            public void onFailure(Method method, Throwable exception) {
+                if (exception.getCause() instanceof JSONException) {
+                    // Only for backward compatbility test
+                    finishTest();
+                }
+
+                fail(exception.getMessage());
+            }
+        });
+    }
+
+}

--- a/restygwt/src/test/java/org/fusesource/restygwt/client/complex/StringEncoderDecoderTestGwt.java
+++ b/restygwt/src/test/java/org/fusesource/restygwt/client/complex/StringEncoderDecoderTestGwt.java
@@ -18,7 +18,9 @@
 
 package org.fusesource.restygwt.client.complex;
 
+import javax.ws.rs.Consumes;
 import javax.ws.rs.GET;
+import javax.ws.rs.POST;
 import javax.ws.rs.Path;
 import javax.ws.rs.Produces;
 import javax.ws.rs.core.MediaType;
@@ -48,6 +50,14 @@ public class StringEncoderDecoderTestGwt extends GWTTestCase {
         @GET
         @Produces(MediaType.TEXT_PLAIN)
         String getAsPlainText();
+
+        @POST
+        @Path("set/json")
+        void setAsJson(String text);
+
+        @POST
+        @Path("set/text")
+        void setAsPlainText(String text);
     }
 
     @Path("/strings")
@@ -63,6 +73,16 @@ public class StringEncoderDecoderTestGwt extends GWTTestCase {
         @GET
         @Produces(MediaType.TEXT_PLAIN)
         void getAsPlainText(TextCallback callback);
+
+        @POST
+        @Path("set/json")
+        @Consumes(MediaType.APPLICATION_JSON)
+        void setAsJson(String text, MethodCallback<Void> callback);
+
+        @POST
+        @Path("set/text")
+        @Consumes(MediaType.TEXT_PLAIN)
+        void setAsPlainText(String text, MethodCallback<Void> callback);
     }
 
     public void testJsonString() {
@@ -80,6 +100,31 @@ public class StringEncoderDecoderTestGwt extends GWTTestCase {
             @Override
             public void onFailure(Method method, Throwable exception) {
                 fail(exception.getMessage());
+            }
+        });
+    }
+
+    /**
+     * Test method only success through "onFailure" with restygwt <= 2.0.3 or plain text autodetection set to false (default)
+     */
+    public void testSendJsonString() {
+        StringsAsync strings = GWT.create(StringsAsync.class);
+
+        delayTestFinish(10000);
+
+        strings.setAsJson("\"Json String?\"", new MethodCallback<Void>() {
+            @Override
+            public void onSuccess(Method method, Void response) {
+                fail();
+            }
+
+            @Override
+            public void onFailure(Method method, Throwable exception) {
+                if(400 == method.getResponse().getStatusCode() && "Wrong Format".equals(exception.getMessage())) {
+                    finishTest();
+                } else {
+                    fail();
+                }
             }
         });
     }
@@ -125,6 +170,24 @@ public class StringEncoderDecoderTestGwt extends GWTTestCase {
                 }
 
                 fail(exception.getMessage());
+            }
+        });
+    }
+
+    public void testSendPlainTextString() {
+        StringsAsync strings = GWT.create(StringsAsync.class);
+
+        delayTestFinish(10000);
+
+        strings.setAsPlainText("Plain text String?", new MethodCallback<Void>() {
+            @Override
+            public void onSuccess(Method method, Void response) {
+                finishTest();
+            }
+
+            @Override
+            public void onFailure(Method method, Throwable exception) {
+                fail();
             }
         });
     }

--- a/restygwt/src/test/java/org/fusesource/restygwt/client/complex/StringEncoderDecoderTestGwt.java
+++ b/restygwt/src/test/java/org/fusesource/restygwt/client/complex/StringEncoderDecoderTestGwt.java
@@ -52,11 +52,11 @@ public class StringEncoderDecoderTestGwt extends GWTTestCase {
         String getAsPlainText();
 
         @POST
-        @Path("set/json")
+        @Consumes(MediaType.APPLICATION_JSON)
         void setAsJson(String text);
 
         @POST
-        @Path("set/text")
+        @Consumes(MediaType.TEXT_PLAIN)
         void setAsPlainText(String text);
     }
 
@@ -75,12 +75,10 @@ public class StringEncoderDecoderTestGwt extends GWTTestCase {
         void getAsPlainText(TextCallback callback);
 
         @POST
-        @Path("set/json")
         @Consumes(MediaType.APPLICATION_JSON)
         void setAsJson(String text, MethodCallback<Void> callback);
 
         @POST
-        @Path("set/text")
         @Consumes(MediaType.TEXT_PLAIN)
         void setAsPlainText(String text, MethodCallback<Void> callback);
     }
@@ -120,7 +118,7 @@ public class StringEncoderDecoderTestGwt extends GWTTestCase {
 
             @Override
             public void onFailure(Method method, Throwable exception) {
-                if(400 == method.getResponse().getStatusCode() && "Wrong Format".equals(exception.getMessage())) {
+                if (400 == method.getResponse().getStatusCode() && "Wrong Format".equals(exception.getMessage())) {
                     finishTest();
                 } else {
                     fail();

--- a/restygwt/src/test/java/org/fusesource/restygwt/server/complex/StringEncoderDecoderServlet.java
+++ b/restygwt/src/test/java/org/fusesource/restygwt/server/complex/StringEncoderDecoderServlet.java
@@ -1,0 +1,57 @@
+/**
+ * Copyright (C) 2009-2015 the original author or authors.
+ * See the notice.md file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.fusesource.restygwt.server.complex;
+
+import javax.servlet.ServletConfig;
+import javax.servlet.ServletException;
+
+import org.fusesource.restygwt.client.complex.StringEncoderDecoderTestGwt;
+import org.jboss.resteasy.plugins.server.servlet.HttpServletDispatcher;
+import org.jboss.resteasy.spi.Registry;
+import org.jboss.resteasy.spi.ResteasyProviderFactory;
+
+public class StringEncoderDecoderServlet extends HttpServletDispatcher {
+
+    private static final long serialVersionUID = 1L;
+
+    public static class StringsImpl implements StringEncoderDecoderTestGwt.Strings {
+        @Override
+        public String getAsJson() {
+            return "String as Json";
+        }
+
+        @Override
+        public String getAsPlainText() {
+            return "String as plain text";
+        }
+    }
+
+    @Override
+    public void init(ServletConfig config) throws ServletException {
+        super.init(config);
+
+        final ResteasyProviderFactory providerFactory = (ResteasyProviderFactory) config.getServletContext().getAttribute(ResteasyProviderFactory.class.getName());
+        providerFactory.registerProvider(JsonStringProvider.class);
+
+        final Registry registry = (Registry) config.getServletContext().getAttribute(Registry.class.getName());
+        registry.addPerRequestResource(StringsImpl.class, "/org.fusesource.restygwt.StringEncoderDecoderTestGwt.JUnit");
+        registry.addPerRequestResource(StringsImpl.class, "/org.fusesource.restygwt.AutodetectPlainTextStringEncoderDecoderTestGwt.JUnit");
+    }
+
+}

--- a/restygwt/src/test/java/org/fusesource/restygwt/server/complex/StringEncoderDecoderServlet.java
+++ b/restygwt/src/test/java/org/fusesource/restygwt/server/complex/StringEncoderDecoderServlet.java
@@ -21,13 +21,15 @@ package org.fusesource.restygwt.server.complex;
 import javax.servlet.ServletConfig;
 import javax.servlet.ServletException;
 import javax.ws.rs.Path;
-
 import org.fusesource.restygwt.client.complex.StringEncoderDecoderTestGwt;
 import org.jboss.resteasy.plugins.server.servlet.HttpServletDispatcher;
 import org.jboss.resteasy.spi.BadRequestException;
 import org.jboss.resteasy.spi.Registry;
 import org.jboss.resteasy.spi.ResteasyProviderFactory;
 
+/**
+ * Servlet to check restygwts autodetection of text for @Produces and Json/text for @Consumes
+ */
 public class StringEncoderDecoderServlet extends HttpServletDispatcher {
 
     private static final long serialVersionUID = 1L;
@@ -44,16 +46,22 @@ public class StringEncoderDecoderServlet extends HttpServletDispatcher {
             return "String as plain text";
         }
 
+        /**
+         * Method to test POST with {@code @Consumes(MediaType.APPLICATION_JSON)}.
+         * 
+         * @throws BadRequestException
+         *             For restygwt <= 2.0.3 or plain text autodetection set to {@code false} because it sends every String as plain text
+         */
         @Override
         public void setAsJson(String text) {
-            if(!"\"Json String?\"".equals(text)) {
+            if (!"\"Json String?\"".equals(text)) {
                 throw new BadRequestException("Wrong Format");
             }
         }
 
         @Override
         public void setAsPlainText(String text) {
-            if(!"Plain text String?".equals(text)) {
+            if (!"Plain text String?".equals(text)) {
                 throw new BadRequestException("Wrong Format");
             }
         }

--- a/restygwt/src/test/java/org/fusesource/restygwt/server/complex/StringEncoderDecoderServlet.java
+++ b/restygwt/src/test/java/org/fusesource/restygwt/server/complex/StringEncoderDecoderServlet.java
@@ -20,9 +20,11 @@ package org.fusesource.restygwt.server.complex;
 
 import javax.servlet.ServletConfig;
 import javax.servlet.ServletException;
+import javax.ws.rs.Path;
 
 import org.fusesource.restygwt.client.complex.StringEncoderDecoderTestGwt;
 import org.jboss.resteasy.plugins.server.servlet.HttpServletDispatcher;
+import org.jboss.resteasy.spi.BadRequestException;
 import org.jboss.resteasy.spi.Registry;
 import org.jboss.resteasy.spi.ResteasyProviderFactory;
 
@@ -30,6 +32,7 @@ public class StringEncoderDecoderServlet extends HttpServletDispatcher {
 
     private static final long serialVersionUID = 1L;
 
+    @Path("/strings")
     public static class StringsImpl implements StringEncoderDecoderTestGwt.Strings {
         @Override
         public String getAsJson() {
@@ -39,6 +42,20 @@ public class StringEncoderDecoderServlet extends HttpServletDispatcher {
         @Override
         public String getAsPlainText() {
             return "String as plain text";
+        }
+
+        @Override
+        public void setAsJson(String text) {
+            if(!"\"Json String?\"".equals(text)) {
+                throw new BadRequestException("Wrong Format");
+            }
+        }
+
+        @Override
+        public void setAsPlainText(String text) {
+            if(!"Plain text String?".equals(text)) {
+                throw new BadRequestException("Wrong Format");
+            }
         }
     }
 

